lib/Dialect/HW/HWAttributes.cpp

```diff
@@ -48,6 +51,100 @@ void HWDialect::printAttribute(Attribute attr, DialectAsmPrinter &p) const {
   llvm_unreachable("Unexpected attribute");
 }
 
+//===----------------------------------------------------------------------===//
+// OutputFileAttr
+//===----------------------------------------------------------------------===//
+
+static std::string normalizeFilename(const Twine &directory,
+                                     const Twine &filename) {
+  SmallString<128> fullpath;
+
+  // If the filename is an absolute path, we don't need the directory.
+  if (llvm::sys::path::is_absolute(filename))
+    filename.toVector(fullpath);
+  else
+    llvm::sys::path::append(fullpath, directory, filename);
+
+  // If this is a directory target, we need to ensure it ends with a `/`
+  if (filename.isTriviallyEmpty() && !fullpath.endswith("/"))
+    fullpath += "/";
+
+  // Remove any superfluous path parts.
+  llvm::sys::path::remove_dots(fullpath, /*remove_dot_dot=*/true);
+  return std::string(fullpath);
+}
```

Comment:
Do you know if what this is doing is going to handle symbolic links? I.e., is this going to produce a _canonical_ path or is it something else?

I think we need a canonical path so that these will compare correctly across symlinks and whatever else... If there's an issue here, would something like `llvm::FileCollector::PathCanonicalizer::canonicalize(StringRef srcPath)` do the right thing?

Reply:
Okay, ignore the `PathCanonicalizer` thing. What I think we want here is `llvm::sys::path::real_path`.